cmd/callhome.go

```diff
@@ -0,0 +1,133 @@
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"time"
+
+	"github.com/minio/madmin-go"
+	"github.com/minio/minio/internal/logger"
+)
+
+const (
+	// CallhomeSchemaVersion1 is callhome schema version 1
+	CallhomeSchemaVersion1 = "1"
+
+	// CallhomeSchemaVersion is current callhome schema version.
+	CallhomeSchemaVersion = CallhomeSchemaVersion1
+
+	// CallhomeCycleDefault is the default interval between two callhome cycles (24hrs)
+	CallhomeCycleDefault = 24 * time.Hour
+)
+
+// CallhomeInfo - Contains callhome information
+type CallhomeInfo struct {
+	SchemaVersion string             `json:"schema_version"`
+	AdminInfo     madmin.InfoMessage `json:"admin_info"`
+}
+
+var (
+	enableCallhome            = false
+	callhomeLeaderLockTimeout = newDynamicTimeout(30*time.Second, 10*time.Second)
+	callhomeFreq              = &safeDuration{t: CallhomeCycleDefault}
+)
+
+func updateCallhomeParams(ctx context.Context, objAPI ObjectLayer) {
+	alreadyEnabled := enableCallhome
+	enableCallhome = globalCallhomeConfig.Enable
+	callhomeFreq.Update(globalCallhomeConfig.Frequency)
+	if !alreadyEnabled && enableCallhome {
+		initCallhome(ctx, objAPI)
+	}
+}
+
+// initCallhome will start the callhome task in the background.
+func initCallhome(ctx context.Context, objAPI ObjectLayer) {
+	go func() {
+		r := rand.New(rand.NewSource(time.Now().UnixNano()))
+		// Leader node (that successfully acquires the lock inside runCallhome)
+		// will keep performing the callhome. If the leader goes down for some reason,
+		// the lock will be released and another node will acquire it and take over
+		// because of this loop.
+		for {
+			runCallhome(ctx, objAPI)
+			if !enableCallhome {
+				return
+			}
+
+			// callhome running on a different node.
+			// sleep for some time and try again.
+			duration := time.Duration(r.Float64() * float64(callhomeFreq.Get()))
+			if duration < time.Second {
+				// Make sure to sleep atleast a second to avoid high CPU ticks.
+				duration = time.Second
+			}
+			time.Sleep(duration)
+
+			if !enableCallhome {
+				return
+			}
+		}
+	}()
+}
+
+func runCallhome(pctx context.Context, objAPI ObjectLayer) {
+	// Make sure only 1 callhome is running on the cluster.
+	locker := objAPI.NewNSLock(minioMetaBucket, "callhome/runCallhome.lock")
+	lkctx, err := locker.GetLock(pctx, callhomeLeaderLockTimeout)
+	if err != nil {
+		return
+	}
+	defer locker.Unlock(lkctx.cancel)
+
+	ctx := lkctx.Context()
+	defer lkctx.Cancel()
+
+	callhomeTimer := time.NewTimer(callhomeFreq.Get())
+	defer callhomeTimer.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-callhomeTimer.C:
+			if !enableCallhome {
+				// Stop the processing as callhome got disabled
+				return
+			}
+			// Reset the timer for next cycle.
+			callhomeTimer.Reset(callhomeFreq.Get())
+
+			go performCallhome(ctx)
```

Comment:
we cannot do an uncontrolled function invocation like this inside a timer, if there is an going callhome in progress it should block and complete gracefully. 

```suggestion
			performCallhome(ctx)
			
			// Reset the timer for next cycle.
			callhomeTimer.Reset(callhomeFreq.Get())
```

Reply:
Done